clippy_lints/src/suspicious_arguments.rs

```diff
@@ -0,0 +1,141 @@
+                        let def_for_call = &definition[call_idx];
+                        if reverse_call == &def_for_call.0 {
+                            // This is technically being called twice, but it's being
+                            // deduplicated?
+                            span_lint_and_then(
+                                cx,
+                                SUSPICIOUS_ARGUMENTS,
+                                vec![*call_span, *reverse_call_span],
+                                "these arguments are possibly swapped",
+                                |diag| {
+                                    let second_span = definition[def_idx].1;
+                                    diag.span_note(vec![def_for_call.1, second_span], "the arguments are defined here");
+                                },
+                            );
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl<'tcx> LateLintPass<'tcx> for SuspiciousArguments {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if expr.span.from_expansion() {
+            return;
+        }
+
+        if let ExprKind::Call(f, args) = &expr.kind
+            && let ExprKind::Path(qp) = &f.kind {
+                let hirid = match qp {
+                    QPath::Resolved(_, p) => {
+                        let last_segment = p.segments.last().unwrap().res.unwrap();
+
+                        match last_segment {
+                            def::Res::Def(_, defid) => {
+                                if defid.krate != def_id::LOCAL_CRATE {
+                                    return;
+                                }
+                                let local = def_id::LocalDefId {
+                                    local_def_index: defid.index,
+                                };
+                                let hirid = cx.tcx.hir().local_def_id_to_hir_id(local);
+                                hirid
+                            }
+                            _ => return,
+                        }
+                    }
+                    QPath::TypeRelative(_, _) => {
+                        let res = cx.typeck_results().qpath_res(qp, f.hir_id);
+                        let Some(defid) = res.opt_def_id() else { return };
+                        if defid.krate != def_id::LOCAL_CRATE {
+                            return;
+                        }
+                        let local = def_id::LocalDefId {
+                            local_def_index: defid.index,
+                        };
+                        let hirid = cx.tcx.hir().local_def_id_to_hir_id(local);
+                        hirid
+                    }
+                    QPath::LangItem(_, _, _) => { return; }
+                };
+
+
+                let Some(bodyid) = cx.tcx.hir().maybe_body_owned_by(hirid) else { return; };
```

Comment:
There's a few utilities here & there that can do the heavy lifting for this part (`path_res` being `clippy_utils::path_res`)

```suggestion
        if let ExprKind::Call(f, args) = expr.kind
            && let Some(def_id) = path_res(cx, f).opt_def_id()
            && let Some(node) = cx.tcx.hir().get_if_local(def_id)
            && let Some(body_id) = node.body_id()
        {
```

Reply:
Ah thank you! that's much neater, wasn't aware of that.